### ISSUE TITLE
Bytes vs str incompatibility between manifest and gitignore tools

### ISSUE THREAD

**Issue by marmeladema** [web-platform-tests/wpt#19836]
Hello everyone,

I am trying to improve compatibility for Python3 in servo (see https://github.com/servo/servo/issues/23607) and that leads me to also improve compatibility in wpt too.
I noticed that `tools/manifest` uses `tools/gitignore` to parse `.gitignore` files but the former expects to receive path as `bytes` whereas the latter expects path as `str`.
In Python2 those two types are the same but not in Python3 anymore.

I think we have 3 choices to fix this issue:
1. Change `tools/gitignore` to expect `bytes`
2. Change `tools/manifest` to expect `str`
3. Decode all strings when calling `tools/gitignore` from `tools/manifest`

I think option 1. is the one we should go for, and i am willing to work on it, but i would like your opinion on that matter.

Thank you in advance!

**Comment by jdm** [comment]
cc @gsnedders

**Comment by marmeladema** [comment]
Friendly ping :-)
Has anyone had time to look at this?
Maybe @jgraham ?

**Comment by gsnedders** [comment]
See also #8820. 

I don't think it makes sense to really look at `gitignore` in isolation here.

So there's a few bits that make this complex here:

*  The manifest loads the paths from either `os.listdir` (`str`, practically) or from the existing JSON file (`unicode`/`str`).
 * These paths get compared with gitignores (which are parsed from the file as `str`).
 * We also have the paths in `vcs.GitHasher` which are all `bytes`.
 * All of the other caches (the gitignore cache and the mtime cache) use JSON files to store their data, so again file names are represented as `unicode` here.

The most perf sensitive case is when we have an up-to-date manifest and caches already generated and we're updating it and we have a clean git repo.

In that case, we walk over `os.listdir`'s return values (`bytes`), filter them through the `gitignore` cache (`unicode`), check the mtime cache (`unicode`), read the pathhash (`unicode`), and then we're done (at least until we serialize it to JSON, again `unicode`).

Either we need to decode stuff out of listdir and GitHasher to `unicode` or encode all the JSON caches to `bytes`. To me, when so much of this stream is based on JSON (all Unicode) and when Python 3 makes Unicode paths the default, it seems like the sensible approach is to make the whole thing Unicode.

This means encoding/decoding in some hot loops, but avoids implicit `str`/`unicode` conversions in Python 2 (which slowly add up).

At the moment, I think effectively `gitignore.PathFilter` is:

```python
from six import PY3
from typing import *
    
T = TypeVar('T')

if PY3:
    FSType = str
else:
    FSType = bytes

class PathFilter(object):
    def __init__(self, root, extras=None, cache=None):
        # type: (FSType, Optional[List[str]], Optional[MutableMapping[Text, bool]]) -> None

    def filter(self,
               iterator  # type: Iterable[Tuple[FSType, List[Tuple[FSType, T]], List[Tuple[FSType, T]]]]
               ):
        # type: (...) -> Iterable[Tuple[FSType, List[Tuple[FSType, T]], List[Tuple[FSType, T]]]]

    def __call__(self,
                 iterator  # type: Iterable[Tuple[FSType, List[Tuple[FSType, T]], List[Tuple[FSType, T]]]]
                 ):
        # type: (...) -> Iterable[Tuple[FSType, List[Tuple[FSType, T]], List[Tuple[FSType, T]]]]
```

The fundamental problem here is `__init__`'s `cache` isn't the same type as all the tuple members in `filter` and `__call__`.

The first stage to fixing this has to be to fixing that disparity. I *think* we just want to make `FSType = Text` above, regardless of Python version, which solves that disparity, but it necessitates a number of other changes (and because of Py2's implicit conversion it all gets masked at runtime and somewhat in MyPy through bytes -> unciode promotion), and then fixing everything from there.

That's all a bit rambly, but I think that's my understanding of the situation; let me know if anything is unclear!

**Comment by foolip** [comment]
@Hexcles

**Comment by ExE-Boss** [comment]
This was fixed in commits [`b402397`](https://github.com/web-platform-tests/wpt/commit/b4023970550ce0d454e700fad25e957ad1366ceb#diff-9c76a08d73cbcbaa19c0adedcaa3242e) and [`d4cdf40`](https://github.com/web-platform-tests/wpt/commit/d4cdf4021ab48c66b052fb62d91da1a06c8cbcf7#diff-9c76a08d73cbcbaa19c0adedcaa3242e).

**Comment by foolip** [comment]
I'll go ahead and close. If there's some remaining work, everyone feel free to reopen or file another issue.
